sys/include/net/gnrc/ipv6/ext.h

```diff
@@ -79,6 +53,66 @@ void gnrc_ipv6_ext_demux(gnrc_netif_t *netif,
 gnrc_pktsnip_t *gnrc_ipv6_ext_build(gnrc_pktsnip_t *ipv6, gnrc_pktsnip_t *next,
                                     uint8_t nh, size_t size);
 
+#if     defined(MODULE_GNRC_IPV6_EXT) || defined(DOXYGEN)
+/**
+ * @brief   Processes a packet's payload as hop-by-hop option if @p nh is
+ *          pointing to a value equal to @ref PROTNUM_IPV6_EXT_HOPOPT.
+ *
+ * @note    Should be called by @ref net_gnrc_ipv6 before any further processing
+ *          (even forwarding-related) is done to @p pkt.
+ *
+ * @param[in] pkt       An IPv6 packet in receive order. It assumed that
+ *                      gnrc_pktsnip_t::data points to a hop-by-hop option when
+ *                      @p nh points to value equal to
+ *                      @ref PROTNUM_IPV6_EXT_HOPOPT.
+ * @param[in,out] nh    **In:** The @ref net_protnum of gnrc_pktsnip_t::data of
+ *                      @p pkt. <br />
+ *                      **Out:** If @p nh was pointing to a value equal
+ *                      to @ref PROTNUM_IPV6_EXT_HOPOPT on in and the return
+ *                      value is not NULL it points to the next header field
+ *                      of the processed hop-by-hop option header, if the return
+ *                      value is NULL the value undefined.
+ *                      If @p nh was pointing to a value unequal to @ref
+ *                      PROTNUM_IPV6_EXT_HOPOPT on in the value remains
+ *                      unchanged.
+ *
+ * @return  @p pkt with the hop-by-hop option marked on success.
+ * @return  NULL, if packet.was consumed by the option handling.
+ * @return  NULL, on error. @p pkt is released with EINVAL in that case.
+ */
+gnrc_pktsnip_t *gnrc_ipv6_ext_process_hopopt(gnrc_pktsnip_t *pkt,
+                                             uint8_t *nh);
+
+/**
+ * @brief   Processes a packet's extension headers after a potential initial
+ *          hop-by-hop header.
+ *
+ * @note    Should be called by @ref net_gnrc_ipv6 after @ref
+ *          gnrc_ipv6_ext_process_hopopt(), i.e. the (first) hop-by-hop option
+ *          should already be marked in @p pkt. If a hop-by-hop option is found
+ *          in gnrc_pktsnip_t::data of @p pkt, the function will return an
+ *          error.
+ *
+ * @param[in] pkt   An IPv6 packet in receive order.
+ * @param[in] nh    **In:** The @ref net_protnum of gnrc_pktsnip_t::data of
```

Comment:
I had to re-read and cross-reference this description several times before I (think I) got it. Maybe it could be reworded to something more verbose like

    * @param[in] nh    **In:** The first next header in the extension header
    *                  chain, represented by the @ref net_protnum of the
    *                  gnrc_pktsnip_t::data contained in @p pkt. <br />
    *                  **Out:** The next protocol to be processed (since the
    *                  extension headers are marked now), represented by the
    *                  @ref net_protnum of the gnrc_pktsnip_t::data contained
    *                  in @p pkt.

also, shouldn't it say `@param[in, out] nh` then?

Reply:
(verboseness nitpick may also apply to line 68)

Reply:
```
*                  chain, represented by the @ref net_protnum of the
*                  gnrc_pktsnip_t::data contained in @p pkt. <br />
```

But that's wrong. It's the protocol number of the last parsed header, which isn't in `gnrc_pktsnip_t::data` in `pkt` at this point but in `pkt->next`. But I think such descriptions make it localized confusing as well, since you need to know how packet order works in GNRC.

Reply:
> also, shouldn't it say `@param[in, out] nh` then?

Yepp, sorry will fix.

Reply:
> But that's wrong. It's the protocol number of the last parsed header, which isn't in `gnrc_pktsnip_t::data` in `pkt` at this point but in `pkt->next`. But I think such descriptions make it localized confusing as well, since you need to know how packet order works in GNRC.

Also it isn't true, that `*nh` has to come from the packet either. From a testing perspective it would be totally alright to be some extension header related protocol number. It could just point to `PROTNUM_IPV6_EXT_FRAG` and `pkt->data` will be handled as a fragmentation header.

Reply:
I hope the way I reformulated the comment is now better understandable.